add contributor guide for GenAIComp and GenAIExample (#107)
opea-project/docs: community/CONTRIBUTING.md

## All The Ways To Contribute

> ### Contribute a GenAIComponent

Reviewer: should have a blank here? "Contribute a GenAI Component"?

Author: added

> 1. Navigate to [OPEA GenAIComps](https://github.com/opea-project/GenAIComps) and locate the component folder your integration belongs to. If the microservice type already exists, review the [OPEA microservice API](https://opea-project.github.io/0.9.99/developer-guides/OPEA_API.html#opea-micro-service-api) and follow it in your implementation. Otherwise, if you are contributing a brand new microservice type, you need to define and contribute the API specifications. Please start by submit an RFC to get contributors feedback. 

Reviewer 2: 0.9.99 was a testing "release" of the documentation and is gone now... refer to the /latest/ version instead.
```suggestion
1. Navigate to [OPEA GenAIComps](https://github.com/opea-project/GenAIComps) and locate the component folder your integration belongs to. If the microservice type already exists, review the [OPEA microservice API](https://opea-project.github.io/latest/developer-guides/OPEA_API.html#opea-micro-service-api) and follow it in your implementation. Otherwise, if you are contributing a brand new microservice type, you need to define and contribute the API specifications. Please start by submit an RFC to get contributors feedback. 
```

Author: fixed

  - `requirements.txt`: This file is used by Docker to install the necessary dependencies.
  - `Dockerfile`: Used to generate the service container image. Please follow naming conventions:
    - Dockerfile: `Dockerfile.[vendor]_[hardware]`, vendor and hardware in lower case (i,e Dockerfile.amd_gpu)
    - Docker Image: `opea/[microservice type]-[microservice sub type]-[library name]-[vendor]-[hardware]:latest` all lower case (i,e opea/llm-vllm-intel-hpu, opea/llm-faqgen-tgi-intel-hpu-svc)

  - `tests/[microservices type]/` : contains end-to-end test for microservices Please refer to an example [test_asr_whisper.sh](https://github.com/opea-project/GenAIComps/blob/main/tests/asr/test_asr_whisper.sh). Please follow naming convention:`test_[microservice type]_[microservice sub type]_[library name]_on_[vendor]_[hardware].sh` 
  - `tests/cores/` : contains UT test for core python components (orchestrator, gateway...). Please follow naming convention:`test_[core component].sh` 

  - `README.md`: at minimum it should include: microservice description, build and start microservice commands and curl command with expected output.

> 4. Now you have created all the required files, and validated your service. Last step is to modify the `README.md` at the component level `GenaIComps/comp/microservice type` to list your new component. Now you are ready to file your PR!. In the upcoming release maintainers will push the Docker Image to the Docker Hub. 

Reviewer: at the component level `GenaIComps/comp/microservice type` 

should be 

at the component level `GenaIComps/comps/microservice type`

Author: done

5. After your component has been merged you are likely interested to build an application with it, and perhaps contributing it also to OPEA! so please continue to the [Contribute a GenAI Example](#contribute-a-genai-example) guide.

### Contribute a GenAI Example

> OPEA GenAIExamples examples offer simple deployment, testing, and scalability with Docker compose and Kubernetes. When contributing an example, a Docker Compose deployment is the minimum requirement. However, since OPEA is intended for enterprise applications, supporting Kubernetes deployment is highly encouraged. You can find [examples for Kubernetes deployment](https://github.com/opea-project/GenAIExamples?tab=readme-ov-file#deploy-examples) using manifests, Helms Charts, and the [GenAI Microservices Connector (GMC)](https://github.com/opea-project/GenAIInfra/tree/main/microservices-connector). GMC offers additional enterprise features, such as the ability to dynamically adjust pipelines on Kubernetes (e.g., switching to a different LLM on the fly, adding guardrails), composing pipeleines that include external services hosted in public cloud or on-premisees via URL, and supporting sequential, parallel and conditional flows in the pipelines.

Author: done

- Before stitching together all the microservices to build your application, let's make sure all the required building blocks are available!. You will notice in [ChatQnA Flow Chart](https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/README.md) OPEA uses gateways to handle requests and route them to the corresponding megaservices (unless you have an agent that will otherwise handle the gateway function). If you are just making small changes to the application, like swaping one DB for another, you can reuse the existing Gateway but if you are contributing a completely new application, you will need to add a Gateway class. Navigate to [OPEA GenAIComps Gateway](https://github.com/opea-project/GenAIComps/blob/main/comps/cores/mega/gateway.py) and implement how Gateway should handle requests for your application. Note that Gateway implementation is moving to GenAIExamples in future release. 

- Follow the folder structure in the ChatQA example below:

> ```

Author: @dbkinder I don't know how to get all this indentation in VS code but i think the generated doc looks ok. if you want to push a commit go for it

> ```

Reviewer 3: @dbkinder how about we deal with your help the indentation but get the content merged first.

│   │   │       │   └── chatQnA_xeon.yaml
│   │   │       └── manifest
│   │   │           └── chatqna.yaml     
│   │   └── hpu
│   │       └── gaudi
│   │           ├── gmc
│   │           │   └── chatQnA_gaudi.yaml
│   │           └── manifest
│   │               └── chatqna.yaml
│   ├── amd
│   │   ├── cpu      
│   │   │   ├── gmc
│   │   │   └── manifest
│   │   └── gpu
│   │       ├── gmc
│   │       └── manifest
│   ├── README_gmc.md  # K8s quickstar
│   └── README.md      # quickstart
├── README.md
├── tests
│   ├── test_compose_on_gaudi.sh  #could be more tests for different flavors of the app
│   ├── test_gmc_on_gaudi.sh
│   ├── test_manifest_on_gaudi.sh
└── ui

```
- **File Descriptions**:
  - `chatqna.py`: application definition using microservice, megaservice and gateway. There could be multiple .py in the folder based on slight modification of the example application.
  - `docker_build_image/build.yaml`: builds necessary images pointing to the Dockerfiles in the GenAIComp repository. 
  - `docker_compose/vendor/device/compose.yaml`: defines pipeline for  docker compose deployment. For selectng docker image name please follow the naming convention: 

>       - Docker Image: `opea/[example name]-[feature name]:latest` all lower case (i,e: opea/chatqna, opea/codegen-react-ui)

Author: done

  - `kubernetes/vendor/device/manifests/chatqna.yaml`: used for K8s deployemnt

>   - `kubernetes/vendor/device/gmc/chatqna.yaml`: (optional) used for deployment with GMC

Reviewer: maybe we need a brief introduce on what's the GMC at the first time seeing it.

Author: added intro at the beginning of the section

  - `tests/`: at minimum you need to provide an E2E test with Docker compose. If you are contritbutng K8s manifests and GMC yaml, you should also provide test for those. Please follow naming convention: 
    - Docker compose test: `tests/test_compose_on_[hardware].sh`
    - K8s test: `tests/test_manifest_on_[hardware].sh`
    - K8s with GMC test: `tests/test_gmc_on_[hardware].sh`
  - `ui`: (optional)
  - `assets`: nice to have an application flow diagram

#### Additional steps if your contribution is Hardware Specific

> You will need additional step to configure CI/CD for merging your GenAIComp or GenAIExample. 

Reviewer 3: ```suggestion
You will need additional steps to configure the CI/CD for first testing and then deploying your merged your GenAIComp or GenAIExample. 
```

> - Connect hardware into OPEA GHA as self-hosted runner 

Reviewer 3: ```suggestion
- Connect hardware into OPEA GHA as a self-hosted runner 
```

> You will need additional steps to configure the CI/CD for first testing and then deploying your merged your GenAIComp or GenAIExample. 

Reviewer 3: ```suggestion
You will need additional steps to configure the CI/CD for first testing and then deploying your merged  GenAIComp or GenAIExample. 
```

> - Connect hardware into OPEA GHA as a self-hosted runner 

Reviewer 3: ```suggestion
- Connect hardware into OPEA GitHub Actions (GHA) (https://docs.github.com/en/actions)as a self-hosted runner 
```

> - Contribute new test scripts for the new hardware

Reviewer 3: ```suggestion
- Contribute test scripts for the new hardware
```

- Dockerfile for the Component (i,e `GenAIComp/comps/llm/text-generation/tgi/Dockerfile.[vendor]_[hardware]` )

> - Update image build yaml for new images​

Reviewer 3: ```suggestion
- Update the image build yaml for new images​
```

> - Update CI/CD workflow to identify and deploy new test

Reviewer 3: ```suggestion
- Update the CI/CD workflow to identify and deploy the new test(s).
```

> - Update CI/CD workflow to identify and deploy new test

Reviewer 3: ```suggestion
- Update the CI/CD workflow to identify and deploy new test
```

```
GenAIComps
├── comps
│   ├── agent
│   ├── asr
│   ├── chathistory
│   ├── cores
│   │   ├── mega     #orchestrator, gateway, micro_service class code
│   │   ├── proto    #api protocol
│   │   └── telemetry
│   ├── dataprep
│   ├── embeddings
│   ├── feedback_management
│   ├── finetuning
│   ├── guardrails
│   ├── intent_detection
│   ├── knowledgegraphs
│   ├── llms
│   ├── lvms
│   ├── nginx
│   ├── prompt_registry
│   ├── ragas
│   ├── reranks
│   ├── retrievers
│   ├── tts
│   ├── vectorstores
│   └── web_retrievers
└── tests
    ├── agent
    ├── asr
    ├── chathistory
    ├── cores
    ├── dataprep
    ├── embeddings
    ├── feedback_management
    ├── finetuning
    ├── guardrails
    ├── intent_detection
    ├── llms
    ├── lvms
    ├── nginx
    ├── prompt_registry
    ├── reranks
    ├── retrievers

>     ├── test_finetuning_llm_pretraining.sh

Reviewer 4: these finetuning scripts have already been moved to the finetuning directory, please remove these three lines.

Author: removed 3 scripts from tree

    ├── test_finetuning_rerank.sh
    ├── test_reranks_mosec-neuralspeed.sh
    ├── tts
    ├── vectorstores
    └── web_retrievers
```

2. Follow the folder structure in the TEI embedding component below:

```
GenAIComps
├── comps
│   └── embeddings
│       ├── __init__.py
│       └── tei     #vendor name or serving framework name
│           ├── langchain
│           │   ├── docker_compose_embedding.yaml  #could be multiple for different HW
│           │   ├── Dockerfile        

> │           │   ├── Dockerfile_AMD_GPU         

Reviewer 5: Dockerfile_AMD_GPU -> Dockerfile.amd_gpu

Author: done

> │           │   ├── Dockerfile_Nvidia_GPU      

Reviewer 5: Dockerfile_Nvidia_GPU -> Dockerfile.nvidia_gpu

Author: done

│           │   ├── embedding_tei.py    # definition and registration of microservice
│           │   ├── README.md
│           │   └── requirements.txt
│           └── llama_index
│               └── . . .
├── tests
│   └── embeddings
│       ├── test_embeddings_tei_langchain.sh

> │       ├── test_embeddings_tei_langchain_amd_gpu.sh     

Reviewer 5: test_embeddings_tei_langchain_on_amd_gpu.sh

Author: added _on_ With this is consistent with the tests in examples.

│       └── test_embeddings_tei_llama_index.sh
└── README.md

```
- **File Descriptions**:
  - `embedding_tei.py`: This file defines and registers the microservice. It serves as the entrypoint of the Docker container. Refer to [whisper ASR](https://github.com/opea-project/GenAIComps/tree/main/comps/asr/whisper) for a simple example or [TGI](https://github.com/opea-project/GenAIComps/blob/main/comps/llms/text-generation/tgi/llm.py) for a more complex example that required adapting to OpenAI API.
  - `requirements.txt`: This file is used by Docker to install the necessary dependencies.

>   - `Dockerfile` and optionally `docker_compose_yaml`: These files are used to generate the service container image. Please follow naming conventions:

Reviewer 5: No need `docker_compose_yaml`, this one will be clean up in future release.

Author: removed it from folder tree also

>     - Dockerfile: `Dockerfile_[vendor]_[hardware]`

Reviewer 5: `Dockerfile.[vendor]_[hardware]` vendor and hardware should be in lower-case.

Author: done

>     - Docker Compose: `docker_compose_[microservice type]_[vendor]_[hardware]` all lower case (i,e opea/llm-vllm-intel-hpu, opea/llm-faqgen-tgi-intel-hpu-svc)

Reviewer 5: no this one.

Author: removed docker compose

    - Docker Image: `opea/[microservice type]-[microservice sub type]-[library name]-[vendor]-[hardware]:latest`

>   - `tests/` : test should validate building image, starting service, validating microservie, stoping and destroying the container. Please refer to an example [test_asr_whisper.sh](https://github.com/opea-project/GenAIComps/blob/main/tests/asr/test_asr_whisper.sh). Please follow naming convention:`test_[microservice type]-[microservice sub type]-[library name]-[vendor]-[hardware].sh` 

Reviewer 5: Please also introduce `tests/cores`, this is UT test for core python code, and others are e2e test for microservice.

Author: added

5. After your component has been merged you are likely interested to build an application with it, and perhaps contributing it also to OPEA! so please continue to the "Contribute a GenAIExample" guide

### Contribute a GenAIExample

OPEA GenAIExamples examples offer simple deployment, testing, and scalability with Docker compose and Kubernetes. when contributing an example Docker compose deployment is the minimum requirement, but given OPEA is intended for enterprise applications, support for kubernetes deployment is highly encouraged.

- Navigate to [OPEA GenAIExamples](https://github.com/opea-project/GenAIExamples/tree/main) and check the catalog of examples. If you find one that is very similar to what you are looking for, you can contribute your variation of it to that particular example folder. If you are bringing a completly new application you will need to create a separate example folder. 

> - Before stitching together all the microservices to build your application, let's make sure all the required building blocks are available!. You will notice [ChatQnA Flow Chart]() OPEA uses gateways to handle requests and route them to the corresponding megaservices (unless you have an agent that will otherwise handle the gateway function). If you are just making small changes to the application, like swaping one DB for another, you can reuse the existng Gateway but if you are contributing a completely new application, you will need to add a Gateway class. Navigate to [OPEA GenAIComps Gateway](https://github.com/opea-project/GenAIComps/blob/main/comps/cores/mega/gateway.py) and implement how Gateway should handle requests for your application. Note that Gateway implementation will move to GenAIExamples in future release. 

Reviewer 5: [ChatQnA Flow Chart]()? put a link?

Author: I don't know how to put a link to a mermaid flowchart so i added link to the README that constains it @dbkinder do you know how to do that?

Reviewer 2: Yes I do :)

Author: @dbkinder please add it!

  - `docker_compose/vendor/device/compose.yaml`: defines pipeline for  docker compose deployment. For selectng docker image name please follow the naming convention: 
      - Docker Image: `opea/[example name]-[feature name]:latest` all lower case (i,e: opea/chatqna, opea/codegen-react-ui)
  - `kubernetes/vendor/device/manifests/chatqna.yaml`: used for K8s deployemnt
  - `kubernetes/vendor/device/gmc/chatqna.yaml`: (optional) used for deployment with GMC
  - `tests/`: at minimum you need to provide an E2E test with docker compose. If you are contritbutng K8s manifests and GMC yaml, you should also provide test for those. Please follow naming convention: 
    - docker compose test: `tests/test_compose_on_[hardware].sh`
    - K8s test: `tests/test_manifest_on_[hardware].sh`
    - K8s with GMC test: `tests/test_gmc_on_[hardware].sh`
  - `ui`: (optional)
  - `assets`: nice to have an application flow diagram

#### Additional steps if your contribution is Hardware Specific

You will need additional step to configure CI/CD for merging your GenAIComp or GenAIExample. 
- Connect hardware into OPEA GHA as self-hosted runner 

> - Contribute new test scripts for the new hardware:

Reviewer 5: remove `:` ?

> - Dockerfile for the Component (i,e `GenAIComp/comps/llm/text-generation/tgi/docker/Dockerfile_[vendor]_[hardware]` )

Reviewer 5: no `docker` folder. 
`Dockerfile_[vendor]_[hardware]` --> `Dockerfile.[vendor]_[hardware]`

Author: fix

- Update image build yaml for new images​

> - Update CI/CD workflow to identify and deploy new test

Reviewer 5: We can remove this line, after reorg, new CI/CD structure should adapt to most of the test case.

Author: @chensuyue do I use your email contact or something else?

Reviewer 3: Have reached out for help to create an alias for future use -  to spread the load as the community grows and to ensure coverage when folks are on vacation.

#### Additional steps if your contribution is Hardware Specific

You will need additional steps to configure the CI/CD for first testing and then deploying your merged your GenAIComp or GenAIExample. 

> - Connect hardware into OPEA GitHub Actions (GHA) (https://docs.github.com/en/actions)as a self-hosted runner 

Reviewer 2: ```suggestion
- Connect hardware into OPEA GitHub Actions (GHA) (https://docs.github.com/en/actions) as a self-hosted runner 
```
